Article: Everything Github: Continuous Integration, Deployment and Hosting for your Angular App [WIP]

JohannesHoppe opened the pull request:
FYI, I'm still trying to figure out how everything works.
follow up of PR https://github.com/angular-schule/website-articles/pull/6

Question to @shhdharmen 
In your [example](https://github.com/angular-schule/website-articles/pull/6/files#diff-d652dc167fc8680a5e6b4658a867619eR125), there is no `--repo` option? Did you really managed to make this work in real code? This should not work, because angular-cli-ghpages does only add the token via a simple string/replace on the provided parameter. It shouldn't work with the remote of the repository. [see this code](https://github.com/angular-schule/angular-cli-ghpages/blob/98094397c1f5046e7888dea8c5ee3fdc91962c8c/src/engine/engine.ts#L106)

@EdricChan03
FYI, you might want to get involved as well. I'd appreciate your help. 😃 

I'm also a bit of confused if we're really doing the token thing right at all.
see https://github.com/angular-schule/angular-cli-ghpages/issues/73#issuecomment-568277597

JohannesHoppe commented:
You can read a preview of the article here:
https://github.com/angular-schule/website-articles/blob/everything-github/blog/2019-12-everything-github/README.md

EdricChan03 commented:
I'm currently testing to see if the GitHub installation token works. I'll report back if I have any results.

It seems that as it turns out, the [documentation on the `GITHUB_TOKEN` secret](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/authenticating-with-the-github_token#about-the-github_token-secret) states that:

> The installation access token expires after 60 minutes. GitHub fetches a token for each job, before the job begins.

So it seems that it's safe to use the `GITHUB_TOKEN` even if the token is leaked as it'll automatically expire after an hour.

And from the same [documentation I've linked](https://help.github.com/en/actions/automating-your-workflow-with-github-actions/authenticating-with-the-github_token#permissions-for-the-github_token), it seems that: yes, you can use the `GITHUB_TOKEN` approach instead of a PAT as the `deployments` scope has been granted to the token with read/write.

Here's the full list:


Permission | Access type | Access by forked repos
-- | -- | --
checks | read/write | read
contents | read/write | read
deployments | read/write | read
issues | read/write | read
metadata | read | read
packages | read/write | read
pull requests | read/write | read
repository projects | read/write | read
statuses | read/write | read

(I'm not sure if the above was relevant, but I've added it just in case)

JohannesHoppe commented:
Wow, you're fast! 😄

Thank you so much for your efforts. So that means that the GITHUB_TOKEN now has more rights than before?? Because back then you found out that it doesn't work. (see https://github.com/angular-schule/angular-cli-ghpages/issues/73#issuecomment-527405699) I will give it a try, too! 🚀

shhdharmen commented:
> FYI, I'm still trying to figure out how everything works.
> follow up of PR #6
> 
> Question to @shhdharmen
> In your [example](https://github.com/angular-schule/website-articles/pull/6/files#diff-d652dc167fc8680a5e6b4658a867619eR125), there is no `--repo` option? Did you really managed to make this work in real code? This should not work, because angular-cli-ghpages does only add the token via a simple string/replace on the provided parameter. It shouldn't work with the remote of the repository. [see this code](https://github.com/angular-schule/angular-cli-ghpages/blob/98094397c1f5046e7888dea8c5ee3fdc91962c8c/src/engine/engine.ts#L106)
> 
> @EdricChan03
> FYI, you might want to get involved as well. I'd appreciate your help. smiley
> 
> I'm also a bit of confused if we're really doing the token thing right at all.
> see [angular-schule/angular-cli-ghpages#73 (comment)](https://github.com/angular-schule/angular-cli-ghpages/issues/73#issuecomment-568277597)

@JohannesHoppe Yes, you are right. I have mentioned `--repo` in my work flow file. I will create new pull request for changes in existing doc soon. Thanks for the catch. 😊

JohannesHoppe commented:
FYI https://github.com/actions/checkout/issues/75#issuecomment-568468352

shhdharmen commented:
@JohannesHoppe I checked with `actions/checkout@v2`, it worked fine. I made no changes except changing the version of actions/checkout in workflow file and it worked.

shhdharmen commented:
@JohannesHoppe I also noticed that, in the docs which I wrote for angular-schule, I mentioned this [workflow](https://angular.io/guide/testing#configure-cli-for-ci-testing-in-chrome) from [Angular](https://angular.io)'s website to run e2e tests.

But, to run them successfully on Github Actions, we need to follow Angular CLI's [wiki](https://angular.io/guide/testing#configure-cli-for-ci-testing-in-chrome) doc.

EdricChan03 commented:
@shhdharmen Are you sure that it actually worked? Did you actually check if the site was actually deployed to GitHub Pages? Or it's probably because the repository that you were testing this on was marked as private?

EdricChan03 commented:
@shhdharmen And by the way, the wiki doc you've linked is for Angular CLI v6 and older.

EdricChan03 commented:
@JohannesHoppe Actually, I'd forgotten to mention that the token seems to only work for private repositories. Public repositories won't actually auto-build the newly deployed commit.

shhdharmen commented:
> @shhdharmen Are you sure that it actually worked? Did you actually check if the site was actually deployed to GitHub Pages? Or it's probably because the repository that you were testing this on was marked as private?

Yes, it was private repo.

shhdharmen commented:
> @shhdharmen And by the way, the wiki doc you've linked is for Angular CLI v6 and older.

That's true, but if I don't follow [web driver](https://github.com/angular/angular-cli/wiki/stories-continuous-integration#chromedriver) instructions, e2e test is failing. I have checked it on a private repo.

EdricChan03 commented:
> > @shhdharmen Are you sure that it actually worked? Did you actually check if the site was actually deployed to GitHub Pages? Or it's probably because the repository that you were testing this on was marked as private?
> 
> Yes, it was private repo.

Ah, that explains why it worked. As mentioned in the [GitHub Community post](https://github.community/t5/GitHub-Actions/Github-action-not-triggering-gh-pages-upon-push/m-p/26869) I've linked, some users have commented that deploying to GitHub Pages only works if your repository is marked as private. It doesn't seem to work on public repositories though.

EdricChan03 commented:
> > @shhdharmen And by the way, the wiki doc you've linked is for Angular CLI v6 and older.
> 
> That's true, but if I don't follow [web driver](https://github.com/angular/angular-cli/wiki/stories-continuous-integration#chromedriver) instructions, e2e test is failing. I have checked it on a private repo.

As for that, see angular/angular-cli#16265 for more information.

shhdharmen commented:
> 
> 
> > > @shhdharmen And by the way, the wiki doc you've linked is for Angular CLI v6 and older.
> > 
> > 
> > That's true, but if I don't follow [web driver](https://github.com/angular/angular-cli/wiki/stories-continuous-integration#chromedriver) instructions, e2e test is failing. I have checked it on a private repo.
> 
> As for that, see [angular/angular-cli#16265](https://github.com/angular/angular-cli/issues/16265) for more information.

In that issue, the developer was trying in local machine.

I tried adding `rm -rf node_modules
rm -f package-lock.json` before `npm install`, but it didn't work.

JohannesHoppe commented:
Guys, you're great! Thank you both so much for getting so deeply involved!

I think at the moment there's nothing more we can do than to use good-old the GH_TOKEN - because we want to support public repos. I'll add a new paragraph that addresses the new findings regarding GH_TOKEN. More to come soon.  👍

JohannesHoppe commented:
To improve the overall experience in the article, I made GithHub Actions a first citizen:

https://github.com/angular-schule/angular-cli-ghpages/releases/tag/v0.6.2

JohannesHoppe commented:
Whoa, I think it's done!
@shhdharmen Can you please take a look at this, check for bugs, etc: =)

https://github.com/angular-schule/website-articles/blob/everything-github/blog/2020-01-everything-github/README.md

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+### 1.3. Github Pages
+
+You can use [GitHub Pages](https://pages.github.com/) to host a website directly from a GitHub repository.
+This website can either contain completely static content (i.e. pure HTML files and assets, but no scripts running on the server) oder or you can use Jekyll to convert Markdown files to static content.
```
~~oder~~

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+* unlimited collaborators
+* 3000 total action minutes/month 
+
+Back in 2016, GitHub migrated from a pricing based on the number of repositories to a new shema that considers the number of users. **Legacy plans** do not have access to GitHub Actions.
```
~~shema~~ => schema

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+
+We assume that the majority of our readers have already worked with Angular. But in order for this article to be as complete as possible, we will very shortly set up a simple website based on Angular.
+
+1. If you have not already done so, please install [Node.js](https://nodejs.org), [Google Chrome](https://www.google.com/chrome/) and optionally [Visual Studio Code](https://code.visualstudio.co) with the [Angular-Schule: Extension Pack](https://marketplace.visualstudio.com/items?itemName=angular-schule.angular-schule-extension-pack). Later on we also need [Git](https://git-scm.com/) for the depl ...
```
Verlinkung falsch: ~~https://code.visualstudio.co~~ => https://code.visualstudio.com

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
````diff
@@ -0,0 +1,706 @@
+    </p>
+    ```
+
+    This example loads data from a remote API. Our hope is that later we will be able to findboth the regular text and as well the loaded book in the index of our preferred searchengine. We have also added Bootstrap to make the result look like this:
````
~~findboth~~ => find both
~~searchengine~~ => search engine

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+   
+## 4. A first deployment to GitHub Pages
+
+We are now ready to host our first app on Pages.
```
~on Pages~ => on Github Pages

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+## 4. A first deployment to GitHub Pages
+
+We are now ready to host our first app on Pages.
+For this, the project must be compiled first and the compiled asseds should be pushed to a new branch.
```
~asseds~ => assets

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+3. The app should be available at `https://<username>.github.io/<repositoryname>/` soon.
+   If it is not immediately available, you should first wait a moment.
+
+4. If we now check the results (in our case at https://angular-schule.github.io/everything-github-demo/) we will see a blank page! 😲
```
Verlinkung setzen

JohannesHoppe replied:
I think, that should resolve to a link, i will double-check!

JohannesHoppe replied:
you were right!

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+   ng deploy --base-href=/<repositoryname>/
+
+   # so in our case it must be
+   ng deploy --base-href=/everything-github-demo/
```
I would add a hint like: "don't miss the trailing `/`"

JohannesHoppe replied:
👍

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+
+
+The easiest way to grant access to the repository is to implement this with tokens.
+Tokens can be used instead of a password for Git over HTTPS (*this is want we want to do*),
```
~want we want to do~ => what we want to do

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+The following two are relevant for us:
+
+* **`GH_TOKEN` / Personal access token** – As the name suggests, personal access tokens grant rights that a particular user has.
+    In order to prevent the release of too many rights, the range of features can be limited with so-called scopes.
```
scopes maybe in italic when using it with "so called" the first time? _scopes_

JohannesHoppe replied:
maybe like this?

In order to prevent the release of too many rights, the range of features can be limited with so-called "scopes".

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+
+The following example is additional,
+because we have promised for this article that you do not have to pay a cent.
+But the the usage of a private repos with Github Actions and Github Pages is only possible from the pro plan and higher.
```
double "the"

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+The following example is additional,
+because we have promised for this article that you do not have to pay a cent.
+But the the usage of a private repos with Github Actions and Github Pages is only possible from the pro plan and higher.
+But the big advantage is that we do not have to set up a token in a tiresome way.
```
Sentence before starts also with "but", should be changed

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+
+Finally, we would like to shortly describe how you use your own domain.
+GitHub Pages supports custom domains, by placing a specific file in the root directory of the `gh-pages` branch – and `angular-cli-ghpages` will create that file for you.
+<!-- GitHub also provides [detailed instructions for this](https://help.github.com/en/github/working-with-github-pages/about-custom-domains-and-github-pages). -->
```
the original link should be included

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+Setting up a subdomain isn't very hard and if you already own a main domain, then you don't need to make a new investment.
+In our example our app should be accessible under the domain `everything-github-demo.angular.schule`.
+
+1. First we have to change the DNS for the domain by setting up a so-called CNAME record.
```
CNAME record in italic? As it will be used the first time with "so called"

JohannesHoppe replied:
i don't like italic! ^^

JohannesHoppe replied:
so-called `CNAME` record

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+    That CNAME record should point to a domain that looks like this: `<username>.github.io`
+
+    The `<username>` (or organisation name) is in our case `angular-schule`.
+    As an example, the following screenshot shows the required setting for the DNS provider cloudflare.com:
```
link cloudflare?

JohannesHoppe replied:
Should work, I will double-check

d-koppenhagen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,706 @@
+<!-- **⚠️ Stop!   
+Before you start browsing the web any further:  
+Or journey is not over yet!
+Please checkout this [follow-up article from Danny about the static site generator Scully](https://d-koppenhagen.de/blog/2020-01-angular-scully), which is a perfect fit for Angular on Github Pages.** -->
```
You can uncomment it if you want :P

JohannesHoppe replied:
I would even more favour an article, that is real a follow-up for the readers of this article. But this would need some extra work... 😉

JohannesHoppe commented:
> @d-koppenhagen : Maybe a TL;DR section with a list and links of the single steps would be great

Isn't this already delivered with the TOC?

d-koppenhagen commented:
> > @d-koppenhagen : Maybe a TL;DR section with a list and links of the single steps would be great
> 
> Isn't this already delivered with the TOC?

Not really. I mean a TL;DR more like a checklist. The TOC says more about the content and what's explained inside

shhdharmen commented:
@JohannesHoppe Just to avoid any un-wanted errors for our readers, add a step to commit and push at the end of [4. A first deployment to GitHub Pages](https://github.com/angular-schule/website-articles/blob/everything-github/blog/2020-01-everything-github/README.md#4-a-first-deployment-to-github-pages). Otherwise, their pipeline will fail. 😅

shhdharmen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,713 @@
+His Angular Book uses a new Github repository in every chapter (so there are quite a lot of repos 😅).
+So he developed a solution that makes it as easy as possible to bring an Angular App to Github Pages.
+The project exists since 2016 and if it was started today, it would certainly have a nicer name.
+According to Github, 7000 projects already deploing with angular-cli-ghpages.
```
```suggestion
According to Github, 7000 projects already deplyoing with angular-cli-ghpages.
```

shhdharmen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,713 @@
+        http.get('https://api.angular.schule/book/9783864906466')
+          .subscribe(b => this.book = b);
+      }
+    }
```
Add a line, to avoid errors when `ng lint`ing.
```suggestion
    }
    
```

shhdharmen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,713 @@
+      ],
+      bootstrap: [AppComponent]
+    })
+    export class AppModule { }
```
Add a line, to avoid errors when `ng lint`ing.
```suggestion
    export class AppModule { }
    
```

shhdharmen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,713 @@
+    Please do not try to set a value.
+    This will work for the first time, but the setting is not permanent, because `angular-cli-ghpages` will overwrite it again.
+
+
```
Just to avoid any un-wanted errors for our readers, add a step to commit and push at the end of 4. A first deployment to GitHub Pages. Otherwise, their pipeline will fail. 😅
```suggestion
6. Now, add, commit and push the changes:
    ```sh
    git add .
    git commit -m "build: add angular-cli-gh-pages"
    git push origin master
    ```
```

shhdharmen commented:
I followed your instructions and built a public repo/app. It's working and deploying fine. Links below:

1. Repo - [ shhdharmen / HelloAngularGithubActionsPublic](https://github.com/shhdharmen/HelloAngularGithubActionsPublic)
2. Deploy from workflow commit - [main.yml](https://github.com/shhdharmen/HelloAngularGithubActionsPublic/blob/8d912f430e7cb10c899a7e718249cfa37aab7711/.github/workflows/main.yml)
3. Test and deploy from workflow commit - [test: ci tests config](https://github.com/shhdharmen/HelloAngularGithubActionsPublic/commit/03180e35327f86dfe8f52aeac3f3ad989c042351)
4. Github page url - [https://shhdharmen.github.io/HelloAngularGithubActionsPublic/](https://shhdharmen.github.io/HelloAngularGithubActionsPublic/)

shhdharmen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,713 @@
+4.  **Done!** 🚀.  
+   The action will run by itself the first time.
+
+Next time when you push your changes to Github, Github Actions will run the workflow we created and it will deploy your updated app on Github Pages.
```
```suggestion
This action will run by itself the first time and each time when you push your changes to Github.
```

shhdharmen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,713 @@
+    ![start commit](./screenshot_5_start-commit.png)
+
+4.  **Done!** 🚀.  
+   The action will run by itself the first time.
```
If you do this [change](https://github.com/angular-schule/website-articles/pull/7/files#r364254872), you can remove this line:
```suggestion
```

shhdharmen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,713 @@
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+      run: |
+        npm install
+        npm run ng -- deploy --base-href=/everything-github-demo/ --name="<Your Git Username>" --email=your.mail@example.org
```
Wouldn't we need `--repo` option over here?

JohannesHoppe replied:
That's the new feature of v0.6.2 that I made for this article. The repo is now always discovered from the working directory, so that the token can be always applied! 🚀 

https://github.com/angular-schule/angular-cli-ghpages/commit/c5f7b36c437843f2472a04f1816e331f4fa7d0ff

JohannesHoppe commented:
@shhdharmen I updated our blog, it can now have two authors and two bios. Feel free to add a longer text, there is still very much white space:

<img width="1135" alt="Screen Shot 2020-01-09 at 08 28 52" src="https://user-images.githubusercontent.com/640639/72053391-3b1f2900-32c7-11ea-92c7-c6d967f8cca3.png">

JohannesHoppe commented:
FYI, header image

![everything-github](https://user-images.githubusercontent.com/640639/72058637-b933fd80-32d0-11ea-9f85-3ddeeefdfc02.png)

fmalcher commented:
@JohannesHoppe I want an Angular logo somewhere in the header!
Right side of the black stripe, body of the rocket, helmet of the octocat, …

JohannesHoppe commented:
@fmalcher But the Angular logo is there, and it's on fire! 🔥

fmalcher commented:
aaah, I didn't see this... 🙈

shhdharmen commented:
> @fmalcher But the Angular logo is there, and it's on fire! fire

Even I didn't notice it... 😅

JohannesHoppe commented:
How do you guys like that?

![everything-githubA](https://user-images.githubusercontent.com/640639/72083608-64a97600-3302-11ea-83a9-6ddbc58c785d.png)

fmalcher commented:
> In contrast to former CI/CD monsters it is common with all mentioned providers to express the individual steps in the form of a YAML file and to store this file as part of the source code in the repository.

I know what you mean but I don't get the sentence structure. Should be simplified.
It's not clear at first sight what is the contrast when you talk about a thing all services have in common.

fmalcher commented:
The *GitHub plans* section sounds a lot like advertisement. Maybe add a note that you are not affiliated with GitHub in any way but that you are a passionate GitHub user and that you recommend the service from your experience.

fmalcher commented:
I'd not mention our VScode extension pack. I think this is useless for everyone outside of our workshops.

fmalcher commented:
> Setting up a subdomain isn't very hard and if you already own a main domain

This is a *Second-level domain*. Didn't want to change this so I'll leave this up to you

fmalcher commented:
> https://everything-github-demo.angular.schule – please feel free to try it out!

Does not work

fmalcher commented:
> If you inspect the certificate, you will see that we actually have not used a certificate from LetsEncrypt. Instead, we decided for the proxy mode of Cloudflare and used the free certificate provided by them. The proxy optimizes our website over Cloudflare's Content Delivery Network (CDN) and provides a lot of additional features – but that's another story.

How about simply NOT using this feature for the article then? 😅 This is more confusing than useful here and you could omit this side note completely.

fmalcher commented:
> To create an A record, point your apex domain to the following IP addresses from GitHub:

Mention the yource here as those might change.

shhdharmen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,720 @@
+author2: Dharmen Shah
+mail2: shhdharmen@gmail.com
+twitter2: shhdharmen
+bio2: "Dharmen is a Husband, Techie, Full Stack JavaScript Developer, Foodie, Hindi Pop Music from Ahmedabad in India. In his free-time he participates on the development of angular-cli-ghpages."
```
Let me know how much white space is still there. I will think of additional content if needed.
```suggestion
bio2: "Dharmen is a Full Stack JavaScript Developer. His core skills are JavaScript and CSS. He has worked on Angular and React, both.
He has contributed to angular-cli-ghpages docs and likes to contribute to open-source projects.
He is a foodie and loves listening to music, especially Hindi Pop music. In his free time, he likes to cook. He is from Ahmedabad, India."
```

JohannesHoppe commented:
> > Setting up a subdomain isn't very hard and if you already own a main domain
> 
> This is a _Second-level domain_. Didn't want to change this so I'll leave this up to you

I think the term is correct:
> In the Domain Name System (DNS) hierarchy, a subdomain is a domain that is a part of another (main) domain. ([source](https://en.wikipedia.org/wiki/Subdomain))

Also GitHub calls them subdomains, [see this](https://help.github.com/en/github/working-with-github-pages/managing-a-custom-domain-for-your-github-pages-site#configuring-a-subdomain).

JohannesHoppe commented:
> I'd not mention our VScode extension pack. I think this is useless for everyone outside of our workshops.

Fair point!

> > https://everything-github-demo.angular.schule – please feel free to try it out!
> 
> Does not work

Oops! I added and removed it several times for a nice screenshot! 😅 


> > If you inspect the certificate, you will see that we actually have not used a certificate from LetsEncrypt. Instead, we decided for the proxy mode of Cloudflare and used the free certificate provided by them. The proxy optimizes our website over Cloudflare's Content Delivery Network (CDN) and provides a lot of additional features – but that's another story.
> 
> How about simply NOT using this feature for the article then? 😅 This is more confusing than useful here and you could omit this side note completely.

I agree that this is confusing and I think we need to add a new paragraph for this. One point where Netfly is said to be better is asset optimization:
https://www.netlify.com/github-pages-vs-netlify/

FYI

<img width="1192" alt="Screen Shot 2020-01-10 at 23 50 49" src="https://user-images.githubusercontent.com/640639/72191906-1086ba80-3404-11ea-9413-b96fb45aeaed.png">

![image](https://user-images.githubusercontent.com/640639/72191918-1c727c80-3404-11ea-8377-228851c04440.png)

Fur sure, it's not my job to mention a direct competitor. But still I now have to check if and what Github normally optimizes. There are opposing statements on the net....

fmalcher commented:
> Also GitHub calls them subdomains, see this.

"Subdomain" is fine, I was worried about "main domain", never heard this before. https://en.wikipedia.org/wiki/Second-level_domain

> One point where Netlify is said to be better is asset optimization:

I would not start this discussion at all. In a "everything GitHub" article it's perfectly fine to just show the tools provided by GitHub. Maybe mention that asset optimization is not very good at GitHub and tehrefore it is recommended to use some sort of CDN in front of it.

> I now have to check if and what Github normally optimizes. There are opposing statements on the net....

You don't have to! Just say "Please keep in mind that for further optimizations it is recommended to use a CDN as an additional service in front of the hosted website. E.g. Cloudflare does a good job here."

Or just skip the SSL termination at cloudflare so you don't need to add the part about why the SSL cert is not issued by GitHub but by Cloudflare.

JohannesHoppe commented:
@fmalcher I'm convinced! 😉

shhdharmen commented on blog/2020-01-everything-github/README.md:
```diff
@@ -0,0 +1,719 @@
+author2: Dharmen Shah
+mail2: shhdharmen@gmail.com
+twitter2: shhdharmen
+bio2: "Dharmen is a full stack JavaScript developer from Ahmedabad, India with core skills in JavaScript and CSS. He enjoys working with both Angular and React. Dharmen contributes to some open source projects and has also contributed to the documentation of angular-cli-ghpages documents. By the way, he is a foodie and loves listening to music, especially Hindi Pop music."
```
```suggestion
bio2: "Dharmen is a full stack JavaScript developer from Ahmedabad, India with core skills in JavaScript and CSS. He enjoys working with both Angular and React. Dharmen contributes to some open source projects and has also contributed to the documentation of angular-cli-ghpages. By the way, he is a foodie and loves listening to music, especially Hindi Pop music."
```

JohannesHoppe commented:
Thank you all very much! I will launch the spaceship on Monday! 🚀
